Fix the broken weight multiplier update function

kianenigma opened the pull request:
Closes https://github.com/paritytech/substrate/issues/6328
polkadot companion: https://github.com/paritytech/polkadot/pull/1263

The weight to multiplier update logic was broken. Here's why: https://github.com/paritytech/substrate/issues/6328

This PR will

1. Fix the broken calculation. 
2. Removes the `+1` (`multiply_accumulate`) logic when applying the multiplier. Henceforth, the multiplier will be only positive in `[0, inf]`. 
3. Fix the broken assumption that `TargetBlockFullness` (25%) is independent of `AvailableBlockRatio`(75%). **NO**. The research page clearly states 25% of the normal quota, hence, we now do `target = AvailableBlockRatio * TargetBlockFullness`. 
4. Makes `V` configurable. This dictates how fast the fee changes. Also, I've added lots of documents about what it means and how it could be interpreted. Once I draft a polkadot PR, V should be `0.00003` in polkadot since it has 6 sec block time. Here, since we have 3 seconds, I made it equal to `0.00001`. 

A critical issue is that in the fixed formula, we can never reach absolute zero, otherwise, we will never recover from it. Arguably, if we start from `1` and follow the formula, even on infinite empty blocks, we will never reach absolute zero in the formula if we had infinite precision. But with our limited `Fixed` types, we surely do, and once we do, we need to make sure we can recover from this. 

For this I propose a lower cap. Note that this cap cannot be any `0 + epsilon` value. If, for example, lower cap is `1/10^18` and we reach it, even a full block cannot increase it ever again, due to the same limitation of `Fixed`. So it needs to be a small, but sufficiently large value to ensure _recoverability_.  

I've thus added a sanity check to make sure that for any constant that we chose, a block that has 101/100 of the target weight can increase it:

```rust 
#[test]
fn multiplier_can_grow_from_zero() {
	// if the min is too small, then this will not change, and we are doomed forever.
	// the weight is 1% bigger than target.
	run_with_system_weight(target() * 101 / 100, || {
		let next = runtime_multiplier_update(MIN_MULTIPLIER);
		assert!(next > MIN_MULTIPLIER, "{:?} !>= {:?}", next, MIN_MULTIPLIER);
	})
}
```

But still, this is open to suggestion.

- [x] This const is currently a const in the impl file. Perhaps I can move it to the runtime so that it can be configurable. 
- [x] ~~Now might be a good time to consider adding an upper bound as well. Although, I am not sure if it is needed. If the chain is soooo busy that it is always growing, perhaps the fee should escalate so that the traffic decreases~~. 
- [x] We must now move the type of the Multiplier from FixedI128 to FixedU128. I reckon this will need a micro runtime migration since the encodings are different. 
- [x] Potentially a follow up: Move this implementation to transaction payment module and instantiate in the runtime. It only needs the system module, which we have access to everywhere. I hate the fact that we have multiplier copies of it here and in polkadot. 
  - [ ] I've done this, but a minor problem now is that the tests for the multiplier live in the runtime top level lib file, because most of them depend on the configurations of the struct. Nonetheless, we should move some to the transaction payment module.

kianenigma commented on bin/node/runtime/src/impls.rs:
```diff
@@ -47,21 +47,71 @@ impl Convert<u128, Balance> for CurrencyToVoteHandler {
+/// - `v ≤ p / k * (s − s')`
+/// - or, solving for `p`: `p >= v * k * (s - s')`
+///
+/// where `p` is the amount of change of `k` blocks.
```
```suggestion
/// where `p` is the amount of change over `k` blocks.
```

kianenigma commented on bin/node/runtime/src/impls.rs:
```diff
@@ -47,21 +47,71 @@ impl Convert<u128, Balance> for CurrencyToVoteHandler {
+/// p >= 0.00001 * 28800 * 3 / 4
+/// p >= 0.216
+///
+/// Meaning that fees can change at most by ~20% per day, given extreme congestion.
```
```suggestion
/// Meaning that fees can change by around ~21% per day, given extreme congestion.
```

kianenigma commented on bin/node/runtime/src/impls.rs:
```diff
@@ -47,21 +47,71 @@ impl Convert<u128, Balance> for CurrencyToVoteHandler {
+/// https://w3f-research.readthedocs.io/en/latest/polkadot/Token%20Economics.html
+pub struct TargetedFeeAdjustment<S, V>(sp_std::marker::PhantomData<(S, V)>);
+
+// TODO: migrate FixedI128 to FixedU128.
+// TODO: think about how this value + how to store/configure it.
```
```suggestion
```

kianenigma commented on bin/node/runtime/src/impls.rs:
```diff
@@ -142,49 +197,83 @@ mod tests {
+	fn max_change_per_day() {
+		run_with_system_weight(max(), || {
+			let mut fm = Multiplier::one();
+			// See the example in the module doc. Changes can be at most 0.216, hence fm < 1.216
```
I think this is wrong, need to double check. Changes need to be at **least** `0.216`.

kianenigma commented on primitives/arithmetic/src/fixed_point.rs:
```diff
@@ -372,6 +372,23 @@ macro_rules! implement_fixed {
+
+			#[cfg(any(feature = "std", test))]
+			pub fn from_fraction(x: f64) -> Self {
+				Self((x * (<Self as FixedPointNumber>::DIV as f64)) as $inner_type)
```
I am confused why when I used `$div` this did not work 🤔 
```suggestion
				Self((x * ($div as f64)) as $inner_type)
```

kianenigma commented:
Code changes are in a way that I really recommend reviewing with the split view on github.

athei commented on frame/transaction-payment/src/lib.rs:
```diff
@@ -199,8 +199,8 @@ impl<T: Trait> Module<T> where
-			let targeted_fee_adjustment = NextFeeMultiplier::get();
-			let adjusted_fee = targeted_fee_adjustment.saturating_mul_acc_int(adjustable_fee);
+			let targeted_fee_adjustment = Self::next_fee_multiplier();
+			let adjusted_fee = targeted_fee_adjustment.saturating_mul_int(adjustable_fee);
```
This seems wrong. Now we have
```
adjusted_fee = targeted_fee_adjustment * adjustable_fee
```

instead of 
```
adjusted_fee = adjustable_fee + targeted_fee_adjustment * adjustable_fee
```

shawntabrizi replied:
In his PR notes:

> Removes the +1 (multiply_accumulate) logic when applying the multiplier. Henceforth, the multiplier will be only positive in [0, inf].

athei replied:
I see. So this is intended. Is it really necessary to do this? Cause that brakes off-chain calculation and we need to support all versions of the calculation by switching on the runtime versions. I am not saying that this is not possible just that it would be preferable for off-chain calculation if this could be fixed just by changing how the `Multiplier` updates.

shawntabrizi replied:
My 2 cents: I totally feel for the pain of breaking changes and backwards compatibility. That being said, its my personal opinion that the (n+1) is super confusing, and knowing that this change will not only effect us, but all the future chains building on substrate, I think it is best to nip this in the bud while we are still early in the ecosystem

kianenigma replied:
I agree with what Shawn, I rather break things now. The more we wait, more chains will build upon the old system and at some point it will be beyond fixing.

athei replied:
It's fine if you feel strongly about that. Just wanted to bring that up.

kianenigma commented:
Fixed a bunch more tests, added migration, clean some more code up. Should be ready to review. 

@athei would be good to know what you think of the calculation
@shawntabrizi same about the micro-migration. I was tempted to do it without a release enum to have smaller footprint.

kianenigma commented on bin/node/runtime/src/impls.rs:
```diff
@@ -47,89 +45,63 @@ impl Convert<u128, Balance> for CurrencyToVoteHandler {
-
 #[cfg(test)]
-mod tests {
+mod multiplier_tests {
```
indeed, see my final todo item up top: 

> I've done this, but a minor problem now is that the tests for the multiplier live in the runtime top level lib file, because most of them depend on the configurations of the struct. Nonetheless, we should move some to the transaction payment module.

thiolliere commented on frame/transaction-payment/src/lib.rs:
```diff
@@ -103,6 +217,38 @@ decl_module! {
 			});
 		}
+
+		fn on_runtime_upgrade() -> Weight {
```
as the computation of fee changed, is the current Multiplier a good value for new computation mecanism ?

kianenigma replied:
So the migration should do the following: The old multiplier is read. if it is negative it will be set to zero. Else, it will be set to its old value in FixedU128 (we could improve this by setting it to `old + 1` actually. But it doesn't matter much). 

We know that in our chains including Kusama the current value is `-1` so it will be read from storage, then it will set to 0, then the first `convert()` will do a `.max(min_multiplier)` on this which will set it to a small number defined in the companion PR.

thiolliere replied:
ok so we are good starting with min_multiplier for our chains

shawntabrizi commented on frame/transaction-payment/src/lib.rs:
```diff
@@ -103,6 +217,38 @@ decl_module! {
+			type OldInner = <OldMultiplier as FixedPointNumber>::Inner;
+			type Inner = <Multiplier as FixedPointNumber>::Inner;
+
+			if let Releases::V1Ancient = StorageVersion::get() {
```
this is always one storage read

kianenigma commented on bin/node/executor/tests/basic.rs:
```diff
@@ -52,16 +49,17 @@ use self::common::{*, sign};
-	base_fee + fee_multiplier.saturating_mul_acc_int(length_fee + weight_fee)
+/// Default transfer fee. This will use the same logic is implemented in transaction-payment module.
+///
+/// Note that uses reads the multiplier from storage directly, hence to get the fee of `extrinsic`
```
```suggestion
/// Note that reads the multiplier from storage directly, hence to get the fee of `extrinsic`
```

apopiak commented on bin/node/executor/tests/basic.rs:
```diff
@@ -52,16 +49,17 @@ use self::common::{*, sign};
-	let weight_fee = <Runtime as pallet_transaction_payment::Trait>::WeightToFee::calc(&weight);
-
-	base_fee + fee_multiplier.saturating_mul_acc_int(length_fee + weight_fee)
+/// Default transfer fee. This will use the same logic is implemented in transaction-payment module.
```
```suggestion
/// Default transfer fee. This will use the same logic that is implemented in transaction-payment module.
```

apopiak commented on frame/transaction-payment/src/lib.rs:
```diff
@@ -57,13 +57,125 @@ use sp_runtime::{
+/// where `p` is the amount of change over `k` blocks.
+///
+/// Hence:
+/// - in an fully congested chain: `p >= v * k * (1 - s')`.
```
```suggestion
/// - in a fully congested chain: `p >= v * k * (1 - s')`.
```

apopiak commented on frame/transaction-payment/src/lib.rs:
```diff
@@ -57,13 +57,125 @@ use sp_runtime::{
+/// - in an fully congested chain: `p >= v * k * (1 - s')`.
+/// - in an empty chain: `p >= v * k * (-s')`.
+///
+/// For example, when all block are full, and 28800 blocks per day (default in `substrate-node`)
```
```suggestion
/// For example, when all blocks are full and there are 28800 blocks per day (default in `substrate-node`)
```

apopiak commented on frame/transaction-payment/src/lib.rs:
```diff
@@ -57,13 +57,125 @@ use sp_runtime::{
+/// p >= 0.00001 * 28800 * 0.8125
+/// p >= 0.234
+///
+/// Meaning that fees can change by around ~21% per day, given extreme congestion.
```
```suggestion
/// Meaning that fees can change by around ~23% per day, given extreme congestion.
```

kianenigma commented on bin/node/runtime/src/impls.rs:
```diff
@@ -142,49 +114,84 @@ mod tests {
-			let mut fm = Multiplier::default();
+	fn multiplier_can_grow_from_zero() {
+		// if the min is too small, then this will not change, and we are doomed forever.
+		// the weight is 1/10th bigger than target.
```
```suggestion
		// the weight is 1/100th bigger than target.
```

this is a very important test!

athei commented on frame/system/src/lib.rs:
```diff
@@ -570,7 +570,7 @@ decl_module! {
 		// TODO: This should only be available for testing, rather than in general usage, but
 		// that's not possible at present (since it's within the decl_module macro).
-		#[weight = (*_ratio * T::MaximumBlockWeight::get(), DispatchClass::Operational)]
+		#[weight = *_ratio * T::MaximumBlockWeight::get()]
```
Is this an intentional change? Moving from `Operational` to `Normal`?

kianenigma replied:
Yes, this is only used for a test and the test would require it this way. I also have an issue to remove this all-together soon https://github.com/paritytech/substrate/issues/6353

kianenigma commented on frame/transaction-payment/src/lib.rs:
```diff
@@ -576,6 +724,49 @@ mod tests {
 	}
 
+	#[test]
+	fn multiplier_can_fit_all_weight() {
```
this is pretty stupid of me to put here... @thiolliere this should be moved to your integrity checks or similar.

kianenigma replied:
Fixed

athei commented on frame/transaction-payment/src/lib.rs:
```diff
@@ -194,13 +353,14 @@ impl<T: Trait> Module<T> where
-			let adjustable_fee = len_fee.saturating_add(unadjusted_weight_fee);
-			let targeted_fee_adjustment = NextFeeMultiplier::get();
-			let adjusted_fee = targeted_fee_adjustment.saturating_mul_acc_int(adjustable_fee);
+			let adjustable_fee = unadjusted_len_fee.saturating_add(unadjusted_weight_fee);
```
Didn't you say that you change it so that the len_fee is not adjusted by the multiplier?

kianenigma replied:
I posted that 20 mins ago :D not pushed yet :D

kianenigma replied:
Now you can review again.

athei replied:
Looks good.

gnunicorn commented:
bot merge

gavofyork commented:
@kianenigma this is `C1-low` - it doesn't impact node maintainers in any way. it's a runtime-specific change.
